### DISCUSSION TITLE
[General] Error in Estimating Gas Fee for Account Abstraction Transaction: "ECDSA: invalid signature"

### DISCUSSION THREAD

**Discussion by gauravl-tevaeralabs** [zkSync-Community-Hub/zksync-developers#160] [General] [answered]
### Environment

Mainnet

### Provide a brief description of the functionality you're trying to implement.

I am trying to estimate the gas fee for an account abstraction transaction (aaTxn) from a custom/session wallet to another wallet. I am encountering an error stating "ECDSA: invalid signature" during the estimation process. The custom/session wallet has been created and session keys have been configured.

### What is the specific issue or error you're encountering?

While estimating the gas fee for an account abstraction transaction with provider.estimateGas(aaTx), an error is thrown with the message "ECDSA: invalid signature."

**Code Snippet:**
```
const aaTx: types.TransactionRequest = {
    from: ACCOUNT_ADDRESS,
    to: sessionWallet.address,
    chainId: (await provider.getNetwork()).chainId,
    nonce: await provider.getTransactionCount(ACCOUNT_ADDRESS),
    type: 113,
    customData: {
      ergsPerPubdata: utils.DEFAULT_GAS_PER_PUBDATA_LIMIT
    } as types.Eip712Meta,
    value: ethers.utils.parseEther("0.0001"),
    data: "0x"
};
const gasLimit = await provider.estimateGas(aaTx);
```

### Can you share the error messages or logs you're receiving, if any?

Error: failed to validate the transaction. reason: ECDSA: invalid signature

### Have you made any recent changes to the contract before encountering this issue?

No

### Are there any external libraries or contracts that your contract interacts with?

No

### Can you provide the relevant portions of your contract code where the issue is occurring?

**JS Code Snippet:**
```
const aaTx: types.TransactionRequest = {
    from: ACCOUNT_ADDRESS,
    to: sessionWallet.address,
    chainId: (await provider.getNetwork()).chainId,
    nonce: await provider.getTransactionCount(ACCOUNT_ADDRESS),
    type: 113,
    customData: {
      ergsPerPubdata: utils.DEFAULT_GAS_PER_PUBDATA_LIMIT
    } as types.Eip712Meta,
    value: ethers.utils.parseEther("0.0001"),
    data: "0x"
};
const gasLimit = await provider.estimateGas(aaTx);
```

### Have you tried to isolate the problem, and if so, what were the results?

From my perspective, there appears to be a potential deadlock situation. It seems the issue may arise due to the 'customSignature' not being included, possibly within the customData. However, to acquire this signature, we may need a transaction with a gas estimation. This could create a possible circular dependency, as the gas estimation seems to require the signature, but to obtain the signature, we may again need the gas estimation.

### What steps have you already taken to try to resolve the issue?

I have reviewed relevant documentation and searched for similar issues in forums and discussion groups to find possible solutions or workarounds that could resolve the deadlock or the 'ECDSA: invalid signature' error, but I haven't found a resolution yet.

I've attempted to refine the transaction object and ensured that other fields are correctly populated, and reviewed the session wallet and session keys configuration to make sure they are set up properly.

**Comment by dutterbutter** [comment]
Hello 👋 

If you have not yet, you can review this part of the[ AA tutorial](https://github.com/matter-labs/tutorials/blob/ed847d9a34677af74f519a548a4b63017f07e18d/custom-aa/deploy/deploy-multisig.ts#L67) which estimates gas prior to fully populating the fields and the proceeds to make use of `customSignature`. 

Feel free to also provide a reproducible repo so we can assist further.

**Reply by gauravl-tevaeralabs** [reply]
Hi @dutterbutter,

Thank you for your response. After reviewing the AA tutorial, I've found a discrepancy in managing transactions. Within the tutorial, a transaction is created using populateTransaction.methodName(). However, when the goal is not to call a specific method but merely transfer funds from a custom wallet to another wallet, this approach falls short.

To illustrate the issue, consider reproducing the circumstances using the same AA tutorial and attempting to transfer funds from a MultiSig wallet to another wallet.

The code provided in the tutorial for transferring funds is as follows:

```
let aaTx = await aaFactory.populateTransaction.deployAccount(
    salt,
    // These are accounts that will own the newly deployed account
    Wallet.createRandom().address,
    Wallet.createRandom().address,
);

const gasLimit = await provider.estimateGas(aaTx);
const gasPrice = await provider.getGasPrice();

aaTx = {
    ...aaTx,
    // Deploy a new account using the multisig
    from: multisigAddress,
    gasLimit: gasLimit,
    gasPrice: gasPrice,
    chainId: (await provider.getNetwork()).chainId,
    nonce: await provider.getTransactionCount(multisigAddress),
    type: 113,
    customData: {
        gasPerPubdata: utils.DEFAULT_GAS_PER_PUBDATA_LIMIT,
    } as types.Eip712Meta,
    value: ethers.BigNumber.from(0),
};
```

Since the multisig account is being deployed for the first time, the transaction is populated using await aaFactory.populateTransaction.deployAccount(). However, for subsequent transfers, the option to populate the transaction with the deployAccount() method is not available. Consequently, attempting the following method for the next transfer fails:
```

const aaTx = {
    from: multisigAddress,
    to: toAddress,
    gasLimit: gasLimit,
    gasPrice: gasPrice,
    chainId: (await provider.getNetwork()).chainId,
    nonce: await provider.getTransactionCount(multisigAddress),
    type: 113,
    customData: {
        gasPerPubdata: utils.DEFAULT_GAS_PER_PUBDATA_LIMIT,
    } as types.Eip712Meta,
    value: ethers.BigNumber.from(1),
};
const gasLimit = await provider.estimateGas(aaTx);
const gasPrice = await provider.getGasPrice();
```

It would be insightful to gain your perspective on this matter and possibly pinpoint a solution or alternative method.

Thank you!

**Reply by mpavlovic-txfusion** [reply]
To avoid this circular dependency we usually don't call `estimateGas()` for simple transfers. You can just set the constant value of 20M for the gasLimit and it should succeed. Here is the example https://github.com/matter-labs/daily-spendlimit-tutorial/blob/ae011456cdf271b2b3a554956e321f98aa9ef6c6/TUTORIAL.md?plain=1#L761

**Reply by gauravl-tevaeralabs** [reply]
Hi @mpavlovic-txfusion,

Thank you for your response! For simple transfers, we're hardcoding to a constant at the moment. But shouldn't there be a more robust solution? How is the gas estimated when a custom wallet (denoted as 'AA Txn') tries to interact with external contracts for operations like placing bids, purchasing, or minting NFTs? Relying on a hardcoded gasLimit, as indicated below, may be problematic in some cases:

```
const aaTx: types.TransactionRequest = {
    from: CUSTOM_ACCOUNT_ADDRESS,
    to: NFT_CONTRACT_ADDRESS,
    chainId: (await provider.getNetwork()).chainId,
    nonce: await provider.getTransactionCount(CUSTOM_ACCOUNT_ADDRESS),
    type: 113,
    customData: {
      ergsPerPubdata: utils.DEFAULT_GAS_PER_PUBDATA_LIMIT
    } as types.Eip712Meta,
    value: 120000000000000, // NFT Price
    data: nftContract.interface.encodeFunctionData("mint()")
};
```

**Reply by mpavlovic-txfusion** [reply, marked as answer]
Hello @gauravl-tevaeralabs,

I can not see that you have hardcoded the gasLimit value (there is a separate `gasLimit` field on the `TransactionRequest`).
Hardcoding the constant value is the approach only for simple transfers without calling a specific contract function as done in this example https://github.com/matter-labs/daily-spendlimit-tutorial/blob/ae011456cdf271b2b3a554956e321f98aa9ef6c6/TUTORIAL.md?plain=1#L772

For calls to external contracts you should be able to use the approach that @dutterbutter suggested, which estimates gas prior to fully populating the fields https://github.com/zkSync-Community-Hub/zkync-developers/discussions/160#discussioncomment-7171745.

**Reply by gauravl-tevaeralabs** [reply]
Hi @mpavlovic-txfusion,

I didn't include hardcoded gas in the thread. However, for our production application, we've used it as a workaround. I was looking for an alternative solution, but it seems that what we've implemented and what you're suggesting are the only options. I'm closing this conversation as well. Thank you!

**Comment by augustinov** [comment]
It seems like the issue may be related to the signature generation or verification process for the account abstraction transaction (aaTx). Double-check that the session wallet's keys are correctly configured and that the signature generation process is functioning as expected.

You may need to explicitly provide the signature (customSignature) for the transaction if it's not automatically included in the customData. Ensure that the signature is generated using the appropriate method and format expected by the provider.

If you're encountering a circular dependency where gas estimation requires the signature and vice versa, consider breaking the dependency loop by obtaining the signature separately or by adjusting the workflow to avoid the dependency.

Additionally, verify that all fields of the transaction object (aaTx) are correctly populated and formatted according to the provider's requirements.

If the issue persists, reach out to the provider's documentation or support channels for further assistance on account abstraction transactions and signature validation.

**Comment by backvomeph** [comment]
Ensure that the session key-pair setup is correctly configured and that the custom/session wallet is properly initialized before estimating gas for the account abstraction transaction (aaTx). Double-check the data being passed in the transaction request and verify that all necessary fields, including signatures, are correctly populated.

**Comment by kennycobbs** [comment]
It seems there might be an issue with the signature generation or application process in your code. Ensure that the customSignature field, containing the ECDSA signature, is correctly included in the transaction request before estimating gas. Double-check the implementation of the signing process and verify that the signature is being generated and appended to the transaction data as expected.
